sandbox/plugins/cryptodirectory/src/main/java/org/opensearch/index/store/CryptoDirectory.java

```diff
@@ -0,0 +1,536 @@
+            storeWrappedKey(dataKeyPair.getEncryptedKey());
+        }
+    }
+
+    private void storeWrappedKey(byte[] wrappedKey) {
+        try (IndexOutput out = super.createOutput("keyfile", new IOContext())) {
+            out.writeInt(wrappedKey.length);
+            out.writeBytes(wrappedKey, 0, wrappedKey.length);
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private byte[] getWrappedKey() {
+        try (IndexInput in = super.openInput("keyfile", new IOContext())) {
+            int size = in.readInt();
+            byte[] ret = new byte[size];
+            in.readBytes(ret, 0, ret.length);
+            return ret;
+        } catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param source the file to be renamed
+     * @param dest the new file name
+     */
+    @Override
+    public void rename(String source, String dest) throws IOException {
+        super.rename(source, dest);
+        if (!(source.contains("segments_") || source.endsWith(".si"))) ivMap.put(
+            getDirectory() + "/" + dest,
+            ivMap.remove(getDirectory() + "/" + source)
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for reading
+     * @param context the IO context
+     */
+    @Override
+    public IndexInput openInput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.openInput(name, context);
+        ensureOpen();
+        ensureCanRead(name);
+        Path path = getDirectory().resolve(name);
+        FileChannel fc = FileChannel.open(path, StandardOpenOption.READ);
+        boolean success = false;
+        try {
+            Cipher cipher = CipherFactory.getCipher(provider);
+            String ivEntry = ivMap.get(getDirectory() + "/" + name);
+            if (ivEntry == null) throw new IOException("failed to open file. " + name);
+            byte[] iv = Base64.getDecoder().decode(ivEntry);
+            CipherFactory.initCipher(cipher, this, Optional.of(iv), Cipher.DECRYPT_MODE, 0);
+            final IndexInput indexInput;
+            indexInput = new CryptoBufferedIndexInput("CryptoBufferedIndexInput(path=\"" + path + "\")", fc, context, cipher, this);
+            success = true;
+            return indexInput;
+        } finally {
+            if (success == false) {
+                IOUtils.closeWhileHandlingException(fc);
+            }
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     * @param name the name of the file to be opened for writing
+     * @param context the IO context
+     */
+    @Override
+    public IndexOutput createOutput(String name, IOContext context) throws IOException {
+        if (name.contains("segments_") || name.endsWith(".si")) return super.createOutput(name, context);
```

Comment:
we should also skip ```write.lock``` and ```.del``` ? 

https://lucene.apache.org/core/3_0_3/fileformats.html#Lock%20File
https://lucene.apache.org/core/3_0_3/fileformats.html#Deleted%20Documents

Reply:
I think we are also encrypting ```ivmap``` ?  i think we should be skip  encrypting ivMap.
we will need a integration test to test ```ivMap``` is been loaded correctly post a restart.